File: app/src/main/res/layout/ui/admin/StatisticalSubjectActivity.java
package com.fh.app_student_management.ui.admin;

import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.fh.app_student_management.R;
import com.fh.app_student_management.adapters.admin.SubjectStatisticalRecycleViewAdapter;
import com.fh.app_student_management.data.AppDatabase;
import com.fh.app_student_management.data.entities.Semester;
import com.fh.app_student_management.data.relations.StatisticalOfSubject;
import com.fh.app_student_management.utilities.Utils;

import java.util.ArrayList;
import java.util.List;

public class StatisticalSubjectActivity extends AppCompatActivity {

    private ArrayList<Semester> semesters;
    private ArrayList<String> semesterNames;
    private long selectedSemesterId;
    private String selectedSemesterName;

    private ImageView btnBack;
    private EditText edtSemester;
    private LinearLayout titleTable;
    private TextView txtSemesterName;
    private TextView txtSubjectCount;
    private RecyclerView rvSubject;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.admin_activity_statistical_subject);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        initStatisticalSubjectView();
        handleEventListener();
    }

    private void initStatisticalSubjectView() {
        btnBack = findViewById(R.id.btnBack);
        edtSemester = findViewById(R.id.edtSemester);
        titleTable = findViewById(R.id.titleTable);
        txtSemesterName = findViewById(R.id.txtSemesterName);
        txtSubjectCount = findViewById(R.id.txtSubjectCount);
        rvSubject = findViewById(R.id.rvSubject);

        titleTable.setVisibility(View.GONE);

        semesters = new ArrayList<>(AppDatabase.getInstance(this).semesterDAO().getAll());
        semesterNames = new ArrayList<>(semesters.size() + 1);
        semesterNames.add(0, "--- Chọn học kỳ ---");
        for (int i = 0; i < semesters.size(); i++) {
            String startDate = Utils.formatDate("MM/yyyy").format(semesters.get(i).getStartDate());
            String endDate = Utils.formatDate("MM/yyyy").format(semesters.get(i).getEndDate());
            String semesterName = String.format("%s (%s - %s)", semesters.get(i).getName(), startDate, endDate);
            semesterNames.add(semesterName);
        }
    }

    private void handleEventListener() {
        btnBack.setOnClickListener(v -> finish());

        edtSemester.setOnClickListener(v -> new AlertDialog.Builder(this)
                .setTitle("Chọn học kỳ")
                .setItems(semesterNames.toArray(new CharSequence[0]), (dialog, which) -> {
                    if (which == 0) {
                        selectedSemesterId = 0;
                        edtSemester.setText("");
                        updateStatistical();
                    } else {
                        selectedSemesterId = semesters.get(which - 1).getId();
                        selectedSemesterName = semesterNames.get(which);
                        edtSemester.setText(semesterNames.get(which));
                        updateStatistical();
                    }
                }).show());
    }

    private void updateStatistical() {
        List<StatisticalOfSubject> statisticalOfSubjects = new ArrayList<>();
        if (selectedSemesterId > 0) {
            statisticalOfSubjects = AppDatabase.getInstance(this).statisticalDAO().getStatisticalOfSubject(selectedSemesterId);
        }
        rvSubject.setLayoutManager(new LinearLayoutManager(this));
        rvSubject.setAdapter(new SubjectStatisticalRecycleViewAdapter(this, new ArrayList<>(statisticalOfSubjects)));
        txtSemesterName.setText(selectedSemesterName);
        txtSubjectCount.setText(String.valueOf(statisticalOfSubjects.size()));
        titleTable.setVisibility(View.VISIBLE);
    }
}

File: app/src/main/java/ma/ensa/projet/ui/admin/StatisticalSubjectActivity.kt
package ma.ensa.projet.ui.admin

import ma.ensa.projet.R
import ma.ensa.projet.data.AppDatabase
import ma.ensa.projet.data.entities.Semester
import ma.ensa.projet.utilities.Utils
import android.os.Bundle
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.graphics.Insets
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import ma.ensa.projet.adapters.admin.SubjectStatisticalRecycleViewAdapter
import ma.ensa.projet.data.dto.StatisticalOfSubject


class StatisticalSubjectActivity : AppCompatActivity() {

    private lateinit var semesters: ArrayList<Semester>
    private lateinit var semesterNames: ArrayList<String>
    private var selectedSemesterId: Long = 0
    private var selectedSemesterName: String? = null

    private lateinit var btnBack: ImageView
    private lateinit var edtSemester: EditText
    private lateinit var titleTable: LinearLayout
    private lateinit var txtSemesterName: TextView
    private lateinit var txtSubjectCount: TextView
    private lateinit var rvSubject: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.admin_activity_statistical_subject)

        // Handle window insets
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        initStatisticalSubjectView()
        handleEventListener()
    }



    private fun initStatisticalSubjectView() {
        btnBack = findViewById(R.id.btnBack)
        edtSemester = findViewById(R.id.edtSemester)
        titleTable = findViewById(R.id.titleTable)
        txtSemesterName = findViewById(R.id.txtSemesterName)
        txtSubjectCount = findViewById(R.id.txtSubjectCount)
        rvSubject = findViewById(R.id.rvSubject)

        titleTable.visibility = View.GONE

        // Access database asynchronously
        lifecycleScope.launch {
            // Query database in the IO dispatcher (background thread)
            val semestersList = withContext(Dispatchers.IO) {
                AppDatabase.getInstance(this@StatisticalSubjectActivity)?.semesterDAO()?.getAll()
            }

            // Continue on the main thread with the retrieved data
            semestersList?.let { semesters ->
                this@StatisticalSubjectActivity.semesters = ArrayList(semesters)
                semesterNames = ArrayList(semesters.size + 1)
                semesterNames.add(0, "--- Select Semester ---")

                for (semester in semesters) {
                    val startDate = Utils.formatDate("MM/yyyy").format(semester.startDate)
                    val endDate = Utils.formatDate("MM/yyyy").format(semester.endDate)
                    val semesterName = String.format("%s (%s - %s)", semester.name, startDate, endDate)
                    semesterNames.add(semesterName)
                }
            }
        }
    }

    private fun handleEventListener() {
        btnBack.setOnClickListener { finish() }

        edtSemester.setOnClickListener {
            AlertDialog.Builder(this)
                .setTitle("Select Semester")
                .setItems(semesterNames.toTypedArray()) { dialog, which ->
                    if (which == 0) {
                        selectedSemesterId = 0
                        edtSemester.setText("")
                        updateStatistical()
                    } else {
                        selectedSemesterId = semesters[which - 1].id
                        selectedSemesterName = semesterNames[which]
                        edtSemester.setText(semesterNames[which])
                        updateStatistical()
                    }
                }
                .show()
        }
    }

    private fun updateStatistical() {
        lifecycleScope.launch(Dispatchers.IO) {
            val statisticalOfSubjects: List<StatisticalOfSubject> = if (selectedSemesterId > 0) {
                AppDatabase.getInstance(this@StatisticalSubjectActivity)?.statisticalDAO()?.getStatisticalOfSubject(selectedSemesterId) ?: emptyList()
            } else {
                emptyList()
            }

            // Switch to the main thread to update the UI
            withContext(Dispatchers.Main) {
                rvSubject.layoutManager = LinearLayoutManager(this@StatisticalSubjectActivity)
                rvSubject.adapter = SubjectStatisticalRecycleViewAdapter(this@StatisticalSubjectActivity, ArrayList(statisticalOfSubjects))
                txtSemesterName.text = selectedSemesterName
                txtSubjectCount.text = statisticalOfSubjects.size.toString()
                titleTable.visibility = View.VISIBLE
            }
        }
    }}
